### ISSUE TITLE
C++ kernel module in Semigroups package no longer compiles

### ISSUE THREAD

**Issue by james-d-mitchell** [gap-system/gap#1212]
Since the new build system was merged into the `master` branch of GAP, it is no longer possible to build the kernel module of the Semigroups package. Before this it was working just fine. This can be seen in the latest travis build for Semigroups:

[https://travis-ci.org/gap-packages/Semigroups/jobs/217987376](https://travis-ci.org/gap-packages/Semigroups/jobs/217987376)

This particular error seems to be caused by a missing header include in `list.h` or maybe we are no longer doing the correct thing in the Semigroups package's kernel module.  I did not find any comments about this in PR #1193 or in `README.buildsys.md`, but perhaps I missed something?

If I add `#include "gap.h"` (where `LAST_FN` is defined) in `gap/src/list.h` (I'm guessing this is not the right thing to do), then I get a further torrent of errors: [https://pastebin.com/6fj4ipGL](https://pastebin.com/6fj4ipGL)

I'm using to the following version of clang to produce the file logged above:

```
Apple LLVM version 8.1.0 (clang-802.0.38)
Target: x86_64-apple-darwin16.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

The Semigroups kernel module was initially produced using [https://github.com/fingolfin/PackageMaker](https://github.com/fingolfin/PackageMaker) 
so it is possible that this issue will effect other packages too.

It would be extremely useful to have some guidance on what to do to resolve this.

**Comment by markuspf** [comment]
The problem is that you have a file called `src/gap.h` which conflicts with GAP's own `src/gap.h`.

I am not sure whether there's a good way of solving this without renaming your `gap.h`.

**Comment by james-d-mitchell** [comment]
Ah, ok, that's easy to fix then. Any insight into why this wasn't an issue before?

**Comment by markuspf** [comment]
We changed the handling of includes for the move to a merged HPC-GAP; all gap files now `#include <src/foo.h>` instead of the previous `#include "foo.h"`, and we changed the include search path to be the top of the GAP and HPC-GAP directory.

Now with `#include "foo.h"` the preprocessor looks in the local directory first for the include, with `#include <src/foo.h>` it looks in the search path, so previously it picked the "correct" `gap.h` when  compiling semigroups code, even though there were actually two different `gap.h` around, whereas now it includes your `src/gap.h` and not the one from GAP itself.

(I hope this makes sense, and is actually accurate. Lets see whether @fingolfin has anything else to add)
